src/messages.rs

```diff
@@ -44,97 +15,60 @@ pub enum Inbound {
     PlayerRegistered(structs::PlayerRegistered),
     InvalidPlayerName(structs::InvalidPlayerName),
     HeartBeatResponse(structs::HeartBeatResponse),
-    GameLinkEvent(structs::GameLink),
+    GameLink(structs::GameLink),
+    GameResult(structs::GameResult),
     UnrecognizedMessage
 }
 
-pub fn parse_inbound_msg(msg: &String) -> Result<Inbound, Error> {
-    let msg: Inbound =
-        if msg.contains(GAME_ENDED) {
-            Inbound::GameEnded(try!(from_str(msg)))
-        } else if msg.contains(TOURNAMENT_ENDED) {
-            Inbound::TournamentEnded(try!(from_str(msg)))
-        } else if msg.contains(MAP_UPDATE) {
-            Inbound::MapUpdate(try!(from_str(msg)))
-        } else if msg.contains(SNAKE_DEAD) {
-            Inbound::SnakeDead(try!(from_str(msg)))
-        } else if msg.contains(GAME_STARTING) {
-            Inbound::GameStarting(try!(from_str(msg)))
-        } else if msg.contains(PLAYER_REGISTERED) {
-            Inbound::PlayerRegistered(try!(from_str(msg)))
-        } else if msg.contains(INVALID_PLAYER_NAME) {
-            Inbound::InvalidPlayerName(try!(from_str(msg)))
-        } else if msg.contains(HEART_BEAT_RESPONSE) {
-            Inbound::HeartBeatResponse(try!(from_str(msg)))
-        } else if msg.contains(GAME_LINK_EVENT) {
-            Inbound::GameLinkEvent(try!(from_str(msg)))
-        } else {
-            Inbound::UnrecognizedMessage
-        };
-
-    Ok(msg)
-}
-
-
-pub fn create_play_registration_msg(name: String) -> Result<String, Error> {
-    to_string(&structs::PlayRegistration {
-        type_: String::from(REGISTER_PLAYER_MESSAGE_TYPE),
-        playerName: name,
-        gameSettings: default_gamesettings()
-    })
+pub fn handle_inbound_msg(s: &String) -> Result<Inbound, Error> {
+    let mut v: Value = from_str(s).expect("That is not something I can parse to JSON.");
+    let m            = v.as_object_mut().expect("I expected an object. Got something else.");
+    let typ          = m.remove("type").expect("This object doesn't contain the key `type`.")
+                        .as_str().expect("I expected a string. That was not a string.")
+                        .rsplit('.').next().expect("I expected a string containing at least one dot. This string didn't.")
```

Comment:
I think a better wording here is one that indicates what went wrong, i.e.:
`Unexpected format for the message's type, expected at least one dot. Value was: {:?}`. The same goes for all of the strings in the expect here, they should be reworded to avoid using `I` and instead give an illustrative error for when they eventually get thrown in the user's face.

Otherwise, as I said... God, I wish I thought of this earlier. I spent so many hours making JSON parsing work...